sycl/source/detail/scheduler/commands.cpp

```diff
@@ -1141,21 +1141,12 @@ cl_int MemCpyCommand::enqueueImp() {
 
   auto RawEvents = getPiEvents(EventImpls);
 
-  // Omit copying if mode is discard one.
-  // TODO: Handle this at the graph building time by, for example, creating
-  // empty node instead of memcpy.
-  if (MDstReq.MAccessMode == access::mode::discard_read_write ||
-      MDstReq.MAccessMode == access::mode::discard_write ||
-      MSrcAllocaCmd->getMemAllocation() == MDstAllocaCmd->getMemAllocation()) {
-    Command::waitForEvents(Queue, EventImpls, Event);
-  } else {
-    MemoryManager::copy(
-        MSrcAllocaCmd->getSYCLMemObj(), MSrcAllocaCmd->getMemAllocation(),
-        MSrcQueue, MSrcReq.MDims, MSrcReq.MMemoryRange, MSrcReq.MAccessRange,
-        MSrcReq.MOffset, MSrcReq.MElemSize, MDstAllocaCmd->getMemAllocation(),
-        MQueue, MDstReq.MDims, MDstReq.MMemoryRange, MDstReq.MAccessRange,
-        MDstReq.MOffset, MDstReq.MElemSize, std::move(RawEvents), Event);
-  }
+  MemoryManager::copy(
```

Comment:
> CUDA requires no change. OpenCL will need to be changed in the library itself.

Does it mean that this unconditional copy is expected to slow down OpenCL execution currently?
And why CUDA does not need a change?

Reply:
Maybe my commit comment could be clearer.  There are two sides: unified memory and not.

This particular line of code you are commenting on is for when we do not have unified memory.  In that case, the underlying backend doesn't really matter. SYCL itself is scheduling individual  mem read/writes to maintain the coherency, and this optimization will avoid the needless mem read.

The comment about _CUDA requires no change_  is in the context of when there is unified memory. Then we are scheduling paired map/unmap operations.  Any optimization will have to be performed by the backend (or it's PI interface).  In the case of Level Zero, this PR adds the optimization to its PI interface. In the case of CUDA, it looks like the PI interface is already performing the optimization (and microbenchmarks confirm that).   For OpenCL, the `CL_MAP_WRITE_INVALIDATE_REGION` flag is being passed by the PI to the OpenCL plugin, but no optimization seems to be occurring (as tested by a simple benchmark).

Reply:
> In the case of CUDA, it looks like the PI interface is already performing the optimization (and microbenchmarks confirm that)

Could you double check and spot it in the CUDA PI plugin sources?

> For OpenCL, the CL_MAP_WRITE_INVALIDATE_REGION flag is being passed by the PI to the OpenCL plugin, but no optimization seems to be occurring (as tested by a simple benchmark).

This is strange that OpenCL doesn't optimize this, are you going to follow up with OpenCL team?

Reply:
> Could you double check and spot it in the CUDA PI plugin sources?
It's at https://github.com/intel/llvm/blob/6733c8b0efde2c137e2ad0bea2168dc7fd7591be/sycl/plugins/cuda/pi_cuda.cpp#L4120 .   with a supporting comment, but I never step traced it.  But the benchmark confirms the difference.

> This is strange that OpenCL doesn't optimize this, are you going to follow up with OpenCL team?
Agreed. That is the plan.